# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

import launch
from launch.actions import DeclareLaunchArgument
from launch.substitutions import LaunchConfiguration
from launch_ros.actions import ComposableNodeContainer
from launch_ros.descriptions import ComposableNode


def generate_launch_description():
    launch_args = [
        DeclareLaunchArgument(
            'image_height',
            default_value='576',
            description='The height of the input image'),
        DeclareLaunchArgument(
            'image_width',
            default_value='960',
            description='The width of the input image'),
        DeclareLaunchArgument(
            'featnet_engine_file_path',
            default_value='',
            description='The absolute path to the Bi3D Featnet TensorRT engine plan'),
        DeclareLaunchArgument(
            'segnet_engine_file_path',
            default_value='',
            description='The absolute path to the Bi3D Segnet TensorRT engine plan'),
        DeclareLaunchArgument(
            'max_disparity_values',
            default_value='64',
            description='The maximum number of disparity values given for Bi3D inference'),
    ]

    # Bi3DNode parameters
    image_height = LaunchConfiguration('image_height')
    image_width = LaunchConfiguration('image_width')
    featnet_engine_file_path = LaunchConfiguration('featnet_engine_file_path')
    segnet_engine_file_path = LaunchConfiguration('segnet_engine_file_path')
    max_disparity_values = LaunchConfiguration('max_disparity_values')

    bi3d_node = ComposableNode(
        name='bi3d_node',
        package='isaac_ros_bi3d',
        plugin='nvidia::isaac_ros::bi3d::Bi3DNode',
        parameters=[{
                'image_height': image_height,
                'image_width': image_width,
                'featnet_engine_file_path': featnet_engine_file_path,
                'segnet_engine_file_path': segnet_engine_file_path,
                'max_disparity_values': max_disparity_values}],
        remappings=[('left_image_bi3d', 'rgb_left'),
                    ('right_image_bi3d', 'rgb_right'),
                    ('left_camera_info_bi3d', 'camera_info_left'),
                    ('right_camera_info_bi3d', 'camera_info_right')]
        )

    container = ComposableNodeContainer(
        name='bi3d_container',
        namespace='bi3d',
        package='rclcpp_components',
        executable='component_container_mt',
        composable_node_descriptions=[bi3d_node],
        output='screen'
    )

    final_launch_description = launch_args + [container]
    return (launch.LaunchDescription(final_launch_description))
